---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

```{r}
rm(list=ls())
```


```{r}
library(GEOquery)

GSE89567 = getGEO("GSE89567")

filePaths = getGEOSuppFiles("GSE89567")
```

```{r}
library(Seurat)
library(xlsx)
library(Matrix)
library(future)
library(biomaRt)
library(ggplot2)
library(scds)
library(SingleR)
#plan("multiprocess", workers = 4)
```

```{r}
'mart <- useMart("ENSEMBL_MART_ENSEMBL")
ensembl = useDataset("hsapiens_gene_ensembl",mart=mart)
trl = getBM(attributes=c("ensembl_gene_id","hgnc_symbol", "entrezgene_id", "gene_biotype", "start_position", "end_position", "chromosome_name", "strand", "description","phenotype_description", "mim_morbid_description"), mart= ensembl)
trl$chromosome_name[nchar(trl$chromosome_name)>3] = gsub("_", "", substr(trl$chromosome_name[nchar(trl$chromosome_name)>3], 10, 11))'
```


```{r}
'
setwd(paste(getwd(), "/GSE89567", sep = ""))
meta = as.data.frame(GSE89567$GSE89567_series_matrix.txt.gz)
x = rownames(meta)

X = matrix(sapply(strsplit(as.character(read.csv(strsplit(as.character(meta[x[1],"supplementary_file_1"]),"/")[[1]][9])[,1]), " "), function(X){X[2]}), ncol = 1, dimnames = list(r = sapply(strsplit(as.character(read.csv(strsplit(as.character(meta[x[1],"supplementary_file_1"]),"/")[[1]][9], sep = "\t")[,1]), " "), function(X){X[1]}), c = x[1]))

for(el in x[-1]){
  X = cbind(X,
            (matrix(sapply(strsplit(as.character(read.csv(strsplit(as.character(meta[el,"supplementary_file_1"]),"/")[[1]][9])[,1]), " "), function(X){X[2]}), ncol = 1, dimnames = list(r = sapply(strsplit(as.character(read.csv(strsplit(as.character(meta[el,"supplementary_file_1"]),"/")[[1]][9], sep = "\t")[,1]), " "), function(X){X[1]}), c = el)))[rownames(X),])
}

#save(X, file = "X.Rdata")

'
```


```{r}
meta = as.data.frame(GSE89567$GSE89567_series_matrix.txt.gz)
table(meta $ tumor.ch1)
```


```{r}
X = read.table("GSE89567_IDH_A_processed_data.txt", sep = "\t", header = T)

rownames(X) = X$X
X = X[,-1]

X = as.matrix(X)


X = X[rowSums(X)>0,]
X = X[order(apply(X, 1, var), decreasing = T),]

library(HGNChelper)
x = checkGeneSymbols(rownames(X))

x$ss = sapply(strsplit(x$Suggested.Symbol, " /// "), function(X){X[1]})

#x$ss[is.na(x$ss)] = x[is.na(x$ss),"x"]

X = X[!is.na(x$Suggested.Symbol),]
x = x[!is.na(x$Suggested.Symbol),]

X = X[!duplicated(x$ss),]
x = x[!duplicated(x$ss),]

rownames(X) = x$ss

#x[is.na(x$Suggested.Symbol),]
```
```{r}

```
```{r}
unique(meta$tumor.ch1)
colnames(X)[127:150]
```



```{r}
setwd(paste(getwd(), "/GSE89567", sep = ""))
reference <- BlueprintEncodeData()
scRNA.ds = unique(meta$tumor.ch1)
scRNA.assays = list()
for(el in scRNA.ds){
  x = X[,grepl(el, colnames(X))]
  
  x = CreateSeuratObject(counts = x, project = el, min.cells = 5, min.features = 200)
  x <- PercentageFeatureSet(x, pattern = "^MT-", col.name = "percent.mt")
  doublet = cxds_bcds_hybrid(as.SingleCellExperiment(x), estNdbl = T)
  common <- intersect(rownames(x@assays$RNA@data), rownames(reference))
  singler <- reference[common,]
  singler <- SingleR(test = as.matrix(x@assays$RNA@data[common,]), ref = singler, labels = singler$label.main)
  
  x[["singleR"]] = singler$pruned.labels
  x[["cxds_score"]] = doublet$cxds_score
  x[["bcds_score"]] = doublet$bcds_score
  x[["hybrid_score"]] = doublet$hybrid_score
  x[["Doublet"]] = ifelse(doublet$hybrid_call, "yes", "no")
  #x[["Histology"]] = gsub("genotype/variation: ", "", meta[match(el, meta$name),"X.Sample_characteristics_ch1.2"])
  'x[["Age"]] = age[el]
  x[["IDH"]] = idh[el]
  x[["Tumor"]] = gsub("tissue: ", "", meta[colnames(x),"characteristics_ch1.3"])
  x[["Cell_type"]] = gsub("cell type: ", "", meta[colnames(x),"characteristics_ch1.6"])
  x[["Selection"]] = gsub("	selection: ", "", meta[colnames(x),"characteristics_ch1.8"])'
  #assign(el, x) 
  
  scRNA.assays[[el]] = x   
}
#save(scRNA.assays, file = "scRNA.assays.Rdata")
```

```{r}
#SCT starts here
```

```{r}
scRNA.assays[2]
```
```{r}
for(el in names(scRNA.assays)){
  x = scRNA.assays[[el]]@assays$RNA@counts
  x = as.sparse(10*((2^(x))-1))
  scRNA.assays[[el]]@assays$RNA@counts = x
  scRNA.assays[[el]]@assays$RNA@data = x
  print(el)
}
```


```{r}
library(Matrix)
library(Seurat)
load("scRNA.assays.Rdata")
#scRNA.assays

'for(el in names(scRNA.assays)){
  x = scRNA.assays[[el]]@assays$RNA@counts
  x = as.sparse(round(10*((2^(x))-1)))
  scRNA.assays[[el]]@assays$RNA@counts = x
  scRNA.assays[[el]]@assays$RNA@data = x
  print(el)
}'

scRNA.assays = lapply(scRNA.assays, function(mscRNA){
  mscRNA <- subset(mscRNA, subset = Doublet == "no" & percent.mt < 10 & nFeature_RNA > 400)
  return(mscRNA)
})

scRNA.assays=scRNA.assays[sapply(scRNA.assays, function(x){dim(x@assays$RNA@meta.features)[1]})>10000]
scRNA.assays=scRNA.assays[sapply(scRNA.assays, function(x){length(x$orig.ident)})>200]

for (i in 1:length(scRNA.assays)) {
  scRNA.assays[[i]] <- CellCycleScoring(scRNA.assays[[i]], s.features = cc.genes.updated.2019$s.genes, g2m.features = cc.genes.updated.2019$g2m.genes, set.ident = F)
  scRNA.assays[[i]] <- SCTransform(scRNA.assays[[i]], verbose = T, return.only.var.genes = F) #, vars.to.regress = c("percent.mt", "S.Score", "G2M.Score")
  print(i)
}

library(URD)
D = list()
#p = list()
dim = 50
for (i in 1:length(scRNA.assays)) {
  scRNA.assays[[i]] <- RunPCA(scRNA.assays[[i]], verbose = FALSE, npcs = dim, features = VariableFeatures(scRNA.assays[[i]]))
  
  dims = pcaMarchenkoPastur(M = length(VariableFeatures(scRNA.assays[[i]])), N = dim(scRNA.assays[[i]])[2], pca.sdev = scRNA.assays[[i]]@reductions$pca@stdev, factor = 1)
  dims = (1:dim)[dims]
  
  scRNA.assays[[i]] <- RunUMAP(scRNA.assays[[i]], dims = dims, verbose = FALSE)
  
  scRNA.assays[[i]] <- FindNeighbors(scRNA.assays[[i]], dims = dims, verbose = FALSE)
  scRNA.assays[[i]] <- FindClusters(scRNA.assays[[i]], verbose = FALSE)
  
  D[[i]] = dims
  #p[[i]] = DimPlot(scRNA.assays[[i]], label = TRUE) + NoLegend()
}

gexp = function(scRNA = scRNA, gene, reduction = "UMAP"){
  df = data.frame(scale = scRNA@assays$RNA@scale.data[gene,rownames(scRNA[[]])],
                  counts = scRNA@assays$RNA@counts[gene,rownames(scRNA[[]])],
                  x = scRNA@reductions[[tolower(reduction)]]@cell.embeddings[rownames(scRNA[[]]),1],
                  y = scRNA@reductions[[tolower(reduction)]]@cell.embeddings[rownames(scRNA[[]]),2])
  
  df$value = df$scale
  df$value = df$value + 0.01 - min(df$value[df$counts>0])
  df$value[df$counts==0] = 0
  p = ggplot(df, aes(x = x, y = y, color = value))+
    geom_point() + labs(title = gene, x = paste(reduction, 1, sep = "_"), y = paste(reduction, 2, sep = "_")) +
    scale_color_gradientn(name = "Scaled\nexpression\nlevel",
                          values = c(0, quantile(df$value[df$value>0]))/max(df$value),
                          colors = alpha(c("gray", rev(hcl.colors(5, "RdYlBu"))),0.5))+
    theme_classic()
  return(p)
}

gexp = function(scRNA = scRNA, gene, reduction = "UMAP"){
  df = data.frame(scale = scRNA@assays$SCT@scale.data[gene,rownames(scRNA[[]])],
                  counts = scRNA@assays$RNA@counts[gene,rownames(scRNA[[]])],
                  x = scRNA@reductions[[tolower(reduction)]]@cell.embeddings[rownames(scRNA[[]]),1],
                  y = scRNA@reductions[[tolower(reduction)]]@cell.embeddings[rownames(scRNA[[]]),2])
  
  df$value = df$scale
  df$value = df$value + 0.01 - min(df$value[df$counts>0])
  df$value[df$counts==0] = 0
  p = ggplot(df, aes(x = x, y = y, color = value))+
    geom_point() + labs(title = gene, x = paste(reduction, 1, sep = "_"), y = paste(reduction, 2, sep = "_")) +
    scale_color_gradientn(name = "Scaled\nexpression\nlevel",
                          values = c(0, quantile(df$value[df$value>0]))/max(df$value),
                          colors = alpha(c("gray", rev(hcl.colors(5, "RdYlBu"))),0.5))+
    theme_classic()
  return(p)
}


ig = c( "CD24", "PTPRC", "CLU", "MBP", "GPR17", "SNAP25", "ACTA2", "APOLD1")

for(EL in names(scRNA.assays)){
  x = list()
  x[["clust"]] = DimPlot(scRNA.assays[[EL]], label = TRUE) + NoLegend()
  for(el in ig){
    if(el %in% rownames(scRNA.assays[[EL]])){
      x[[el]] = gexp(scRNA = scRNA.assays[[EL]], gene = el, reduction = "UMAP")
    }
  }
  assign(EL, x)
}
```


```{r}
MGH107
```

```{r}
scRNA.assays$MGH107$prim_ident = ifelse(scRNA.assays$MGH107$seurat_clusters %in% c("1"), "immune",  ifelse(scRNA.assays$MGH107$seurat_clusters %in% c("0"),"glioma", ifelse(scRNA.assays$MGH107@reductions$umap@cell.embeddings[,"UMAP_1"] < (5), "oligodendroglia","astrocyte_d")))

DimPlot(scRNA.assays$MGH107, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH42$prim_ident = ifelse(scRNA.assays$MGH42$seurat_clusters %in% c("0"), "immune",  ifelse(scRNA.assays$MGH42$seurat_clusters %in% c("4"), "astrocyte_d","glioma"))

DimPlot(scRNA.assays$MGH42, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH43$prim_ident = ifelse(scRNA.assays$MGH43$seurat_clusters %in% c("0", "2"), "immune",  ifelse(scRNA.assays$MGH43$seurat_clusters %in% c("5"), "astrocyte_d",ifelse(scRNA.assays$MGH43$seurat_clusters %in% c("4"), "oligodendroglia","glioma")))

DimPlot(scRNA.assays$MGH43, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH44$prim_ident = ifelse(scRNA.assays$MGH44$seurat_clusters %in% c("4"), "immune",  ifelse(scRNA.assays$MGH44$seurat_clusters %in% c("3"), "astrocyte_d","glioma"))

DimPlot(scRNA.assays$MGH44, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH45$prim_ident = ifelse(scRNA.assays$MGH45$seurat_clusters %in% c("1", "4", "7"), "immune",  ifelse(scRNA.assays$MGH45$seurat_clusters %in% c("6"), "oligodendroglia","glioma"))

DimPlot(scRNA.assays$MGH45, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH56$prim_ident = ifelse(scRNA.assays$MGH56$seurat_clusters %in% c("4"), "immune",  ifelse(scRNA.assays$MGH56@reductions$umap@cell.embeddings[,"UMAP_1"] < (-10), "oligodendroglia","glioma"))

DimPlot(scRNA.assays$MGH56, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH57$prim_ident = ifelse(scRNA.assays$MGH57$seurat_clusters %in% c("2"), "astrocyte_d",  "glioma")

DimPlot(scRNA.assays$MGH57, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH61$prim_ident = ifelse(scRNA.assays$MGH61$seurat_clusters %in% c("3"), "immune",  ifelse(scRNA.assays$MGH61$seurat_clusters == "4", "astrocyte_i",  "glioma"))

DimPlot(scRNA.assays$MGH61, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH64$prim_ident = ifelse(scRNA.assays$MGH64$seurat_clusters %in% c("3"), "immune",  ifelse(scRNA.assays$MGH64$seurat_clusters == "4", "astrocyte_d", ifelse(scRNA.assays$MGH64$seurat_clusters == "5", "oligodendroglia", "glioma")))

DimPlot(scRNA.assays$MGH64, group.by = "prim_ident", label = T)
```

```{r}
GSE89567 = scRNA.assays
save(GSE89567, file = "GSE89567_SCT.Rdata")
```

```{r}
#Preparation starts here
```

```{r}
library(Matrix)
library(Seurat)
load("scRNA.assays.Rdata")
scRNA.assays

for(el in names(scRNA.assays)){
  x = scRNA.assays[[el]]@assays$RNA@counts
  x = as.sparse(10*((2^(x))-1))
  scRNA.assays[[el]]@assays$RNA@counts = x
  scRNA.assays[[el]]@assays$RNA@data = x
  print(el)
}

scRNA.assays = lapply(scRNA.assays, function(mscRNA){
  mscRNA <- subset(mscRNA, subset = Doublet == "no" & percent.mt < 10 & nFeature_RNA > 400)
  return(mscRNA)
})

scRNA.assays=scRNA.assays[sapply(scRNA.assays, function(x){dim(x@assays$RNA@meta.features)[1]})>10000]
scRNA.assays=scRNA.assays[sapply(scRNA.assays, function(x){length(x$orig.ident)})>200]

ig = c("ACTA2", "APOLD1", "CD24", "PTPRC", "CLU", "MBP", "GPR17", "SNAP25")
for (i in 1:length(scRNA.assays)) {
  scRNA.assays[[i]] <- NormalizeData(scRNA.assays[[i]])
  scRNA.assays[[i]] <- FindVariableFeatures(scRNA.assays[[i]], selection.method = "vst", nfeatures = 3000)
  scRNA.assays[[i]] <- ScaleData(scRNA.assays[[i]], features = unique(c(ig, VariableFeatures(scRNA.assays[[i]]))))
  
  print(i)
}

library(URD)
D = list()
#p = list()
dim = 50
for (i in 1:length(scRNA.assays)) {
  scRNA.assays[[i]] <- RunPCA(scRNA.assays[[i]], verbose = FALSE, npcs = dim, features = VariableFeatures(scRNA.assays[[i]]))
  
  dims = pcaMarchenkoPastur(M = length(VariableFeatures(scRNA.assays[[i]])), N = dim(scRNA.assays[[i]])[2], pca.sdev = scRNA.assays[[i]]@reductions$pca@stdev, factor = 1)
  dims = (1:dim)[dims]
  
  scRNA.assays[[i]] <- RunUMAP(scRNA.assays[[i]], dims = dims, verbose = FALSE)
  
  scRNA.assays[[i]] <- FindNeighbors(scRNA.assays[[i]], dims = dims, verbose = FALSE)
  scRNA.assays[[i]] <- FindClusters(scRNA.assays[[i]], verbose = FALSE)
  
  D[[i]] = dims
  #p[[i]] = DimPlot(scRNA.assays[[i]], label = TRUE) + NoLegend()
}

gexp = function(scRNA = scRNA, gene, reduction = "UMAP"){
  df = data.frame(scale = scRNA@assays$RNA@scale.data[gene,rownames(scRNA[[]])],
                  counts = scRNA@assays$RNA@counts[gene,rownames(scRNA[[]])],
                  x = scRNA@reductions[[tolower(reduction)]]@cell.embeddings[rownames(scRNA[[]]),1],
                  y = scRNA@reductions[[tolower(reduction)]]@cell.embeddings[rownames(scRNA[[]]),2])
  
  df$value = df$scale
  df$value = df$value + 0.01 - min(df$value[df$counts>0])
  df$value[df$counts==0] = 0
  p = ggplot(df, aes(x = x, y = y, color = value))+
    geom_point() + labs(title = gene, x = paste(reduction, 1, sep = "_"), y = paste(reduction, 2, sep = "_")) +
    scale_color_gradientn(name = "Scaled\nexpression\nlevel",
                          values = c(0, quantile(df$value[df$value>0]))/max(df$value),
                          colors = alpha(c("gray", rev(hcl.colors(5, "RdYlBu"))),0.5))+
    theme_classic()
  return(p)
}


for(EL in names(scRNA.assays)){
  x = list()
  x[["clust"]] = DimPlot(scRNA.assays[[EL]], label = TRUE) + NoLegend()
  for(el in ig){
    if(el %in% rownames(scRNA.assays[[EL]])){
      x[[el]] = gexp(scRNA = scRNA.assays[[EL]], gene = el, reduction = "UMAP")
    }
  }
  assign(EL, x)
}
```

```{r}
MGH42
```

```{r}
scRNA.assays$MGH64$prim_ident = ifelse(scRNA.assays$MGH64$seurat_clusters %in% c("3"), "immune",  ifelse(scRNA.assays$MGH64$seurat_clusters == "4", "astrocyte_d", ifelse(scRNA.assays$MGH64$seurat_clusters == "6", "oligodendroglia", "glioma")))

DimPlot(scRNA.assays$MGH64, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH61$prim_ident = ifelse(scRNA.assays$MGH61@reductions$umap@cell.embeddings[,"UMAP_2"] > (10), "vascular", ifelse(scRNA.assays$MGH61$seurat_clusters == "5", "immune", ifelse(scRNA.assays$MGH61$seurat_clusters == "6", "astrocyte_d", "glioma")))

DimPlot(scRNA.assays$MGH61, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH56$prim_ident = ifelse(scRNA.assays$MGH56$seurat_clusters %in% c("3"), "immune",  ifelse(scRNA.assays$MGH56@reductions$umap@cell.embeddings[,"UMAP_2"] < (-10), "oligodendroglia","glioma"))

DimPlot(scRNA.assays$MGH56, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH57$prim_ident = ifelse(scRNA.assays$MGH57$seurat_clusters %in% c("1"), "astrocyte_d",  "glioma")

DimPlot(scRNA.assays$MGH57, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH107$prim_ident = ifelse(scRNA.assays$MGH107$seurat_clusters %in% c("2"), "immune",  ifelse(scRNA.assays$MGH107@reductions$umap@cell.embeddings[,"UMAP_1"] > 1.2, "oligodendroglia",ifelse(scRNA.assays$MGH107$seurat_clusters == "0", "astrocyte_i", "glioma")))

DimPlot(scRNA.assays$MGH107, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH45$prim_ident = ifelse(scRNA.assays$MGH45$seurat_clusters %in% c("3", "4"), "immune", ifelse(scRNA.assays$MGH45$seurat_clusters == "6", "oligodendroglia", "glioma"))

DimPlot(scRNA.assays$MGH45, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH44$prim_ident = ifelse(scRNA.assays$MGH44$seurat_clusters %in% c("4"), "immune", ifelse(scRNA.assays$MGH44$seurat_clusters == "3", "astrocyte_d", ifelse(scRNA.assays$MGH44@reductions$umap@cell.embeddings[,"UMAP_2"] > 1.5, "oligodendroglia", "glioma")))

DimPlot(scRNA.assays$MGH44, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH43$prim_ident = ifelse(scRNA.assays$MGH43$seurat_clusters %in% c("1", "2"), "immune", ifelse(scRNA.assays$MGH43$seurat_clusters == "4", "astrocyte_d", ifelse(scRNA.assays$MGH43$seurat_clusters == "3", "oligodendroglia", "glioma")))

DimPlot(scRNA.assays$MGH43, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH42$prim_ident = ifelse(scRNA.assays$MGH42$seurat_clusters %in% c("1", "3"), "immune", ifelse(scRNA.assays$MGH42$seurat_clusters == "5", "astrocyte_d", "glioma"))

DimPlot(scRNA.assays$MGH42, group.by = "prim_ident", label = T)
```
```{r}
GSE89567 = scRNA.assays
save(GSE89567, file = "GSE89567.Rdata")
```


```{r}
x <- c(,,,)
#p
```


```{r}
#SCT starts here
```

```{r}
load("scRNA.assays.Rdata")
scRNA.assays
```

```{r}
for(el in names(scRNA.assays)){
  x = scRNA.assays[[el]]@assays$RNA@counts
  x = as.sparse(10*((2^(x))-1))
  scRNA.assays[[el]]@assays$RNA@counts = x
  scRNA.assays[[el]]@assays$RNA@data = x
  print(el)
}

scRNA.assays = lapply(scRNA.assays, function(mscRNA){
  mscRNA <- subset(mscRNA, subset = Doublet == "no" & percent.mt < 10 & nFeature_RNA > 400)
  return(mscRNA)
})

scRNA.assays=scRNA.assays[sapply(scRNA.assays, function(x){dim(x@assays$RNA@meta.features)[1]})>10000]
scRNA.assays=scRNA.assays[sapply(scRNA.assays, function(x){length(x$orig.ident)})>200]
print("***")
scRNA.assays
```



```{r}
```

```{r}
for (i in 1:length(scRNA.assays)) {
  scRNA.assays[[i]] <- CellCycleScoring(scRNA.assays[[i]], s.features = cc.genes.updated.2019$s.genes, g2m.features = cc.genes.updated.2019$g2m.genes, set.ident = F)
  scRNA.assays[[i]] <- SCTransform(scRNA.assays[[i]], verbose = T, return.only.var.genes = F, vars.to.regress = c("percent.mt", "S.Score", "G2M.Score"))
  print(i)
}

library(URD)

D = list()
p = list()
dim = 50
for (i in 1:length(scRNA.assays)) {
  scRNA.assays[[i]] <- RunPCA(scRNA.assays[[i]], verbose = FALSE, npcs = dim)
  
  dims = pcaMarchenkoPastur(M = length(VariableFeatures(scRNA.assays[[i]])), N = dim(scRNA.assays[[i]])[2], pca.sdev = scRNA.assays[[i]]@reductions$pca@stdev, factor = 1)
  dims = (1:dim)[dims]
  
  scRNA.assays[[i]] <- RunUMAP(scRNA.assays[[i]], dims = dims, verbose = FALSE)
  
  scRNA.assays[[i]] <- FindNeighbors(scRNA.assays[[i]], dims = dims, verbose = FALSE)
  scRNA.assays[[i]] <- FindClusters(scRNA.assays[[i]], verbose = FALSE)
  
  D[[i]] = dims
  p[[i]] = DimPlot(scRNA.assays[[i]], label = TRUE) + NoLegend()
}

p
```

```{r}
gexp = function(scRNA = scRNA, gene, reduction = "UMAP"){
  df = data.frame(scale = scRNA@assays$SCT@scale.data[gene,rownames(scRNA[[]])],
                  counts = scRNA@assays$RNA@counts[gene,rownames(scRNA[[]])],
                  x = scRNA@reductions[[tolower(reduction)]]@cell.embeddings[rownames(scRNA[[]]),1],
                  y = scRNA@reductions[[tolower(reduction)]]@cell.embeddings[rownames(scRNA[[]]),2])
  
  df$value = df$scale
  df$value = df$value + 0.01 - min(df$value[df$counts>0])
  df$value[df$counts==0] = 0
  p = ggplot(df, aes(x = x, y = y, color = value))+
    geom_point() + labs(title = gene, x = paste(reduction, 1, sep = "_"), y = paste(reduction, 2, sep = "_")) +
    scale_color_gradientn(name = "Scaled\nexpression\nlevel",
                          values = c(0, quantile(df$value[df$value>0]))/max(df$value),
                          colors = alpha(c("gray", rev(hcl.colors(5, "RdYlBu"))),0.5))+
    theme_classic()
  return(p)
}
```


```{r}
p = list()
pCD45 = list()
pMBP = list()
pCLU = list()
pAPOLD1 = list()
pSNAP25 = list()
pGPR17 = list()
pCD24 = list()
for (i in 1:length(scRNA.assays)) {
  p[[i]] = DimPlot(scRNA.assays[[i]], label = TRUE) + NoLegend()
  pCD45[[i]] = gexp(scRNA = scRNA.assays[[i]], gene = "PTPRC", reduction = "UMAP")
  pMBP[[i]] = gexp(scRNA = scRNA.assays[[i]], gene = "MBP", reduction = "UMAP")
  pCLU[[i]] = gexp(scRNA = scRNA.assays[[i]], gene = "CLU", reduction = "UMAP")
  pAPOLD1[[i]] = gexp(scRNA = scRNA.assays[[i]], gene = "APOLD1", reduction = "UMAP")
  pSNAP25[[i]] = gexp(scRNA = scRNA.assays[[i]], gene = "SNAP25", reduction = "UMAP")
  pGPR17[[i]] = gexp(scRNA = scRNA.assays[[i]], gene = "GPR17", reduction = "UMAP")
  pCD24[[i]] = gexp(scRNA = scRNA.assays[[i]], gene = "CD24", reduction = "UMAP")
  #scRNA.assays[[i]][[]] = scRNA.assays[[i]][[]][,!colnames(scRNA.assays[[i]][[]]) == "prim_ident"]
}
p
```



```{r}
scRNA.assays$MGH42$prim_ident = ifelse(scRNA.assays$MGH42$seurat_clusters == "0", "immune", ifelse(scRNA.assays$MGH42$seurat_clusters == "4", "astrocyte", "glioma"))

DimPlot(scRNA.assays$MGH42, group.by = "prim_ident", label = T)
```



```{r}
scRNA.assays$MGH61$prim_ident = ifelse(scRNA.assays$MGH61$seurat_clusters == "6", "vascular", ifelse(scRNA.assays$MGH61$seurat_clusters == "3", "immune", ifelse(scRNA.assays$MGH61$seurat_clusters == "4", "astrocyte", "glioma")))

DimPlot(scRNA.assays$MGH61, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH43$prim_ident = ifelse(scRNA.assays$MGH43$seurat_clusters %in% c("0", "2"), "immune", ifelse(scRNA.assays$MGH43$seurat_clusters == "4", "oligodendroglia", ifelse(scRNA.assays$MGH43$seurat_clusters == "5", "astrocyte", "glioma")))

DimPlot(scRNA.assays$MGH43, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH45$prim_ident = ifelse(scRNA.assays$MGH45$seurat_clusters %in% c("1", "7", "4"), "immune", ifelse(scRNA.assays$MGH45$seurat_clusters == "6", "oligodendroglia", "glioma"))

DimPlot(scRNA.assays$MGH45, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH44$prim_ident = ifelse(scRNA.assays$MGH44$seurat_clusters %in% c("4"), "immune", ifelse(scRNA.assays$MGH44$seurat_clusters == "3", "astrocyte",  "glioma"))

DimPlot(scRNA.assays$MGH44, group.by = "prim_ident", label = T)
```


```{r}
scRNA.assays$MGH56$prim_ident = ifelse(scRNA.assays$MGH56$seurat_clusters %in% c("4"), "immune", ifelse(scRNA.assays$MGH56@reductions$umap@cell.embeddings[,"UMAP_1"] < (-10), "oligodendroglia",ifelse(scRNA.assays$MGH56$seurat_clusters %in% c("5"), "OPC",   "glioma")))

DimPlot(scRNA.assays$MGH56, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$MGH64$prim_ident = ifelse(scRNA.assays$MGH64$seurat_clusters == "4", "astrocyte", ifelse(scRNA.assays$MGH64$seurat_clusters %in% c("3"), "immune", ifelse(scRNA.assays$MGH64$seurat_clusters == "5", "oligodendroglia",  "glioma")))

DimPlot(scRNA.assays$MGH64, group.by = "prim_ident", label = T)
```




```{r}
scRNA.assays$MGH107$prim_ident = ifelse(scRNA.assays$MGH107$seurat_clusters == "1", "immune", ifelse(scRNA.assays$MGH107$seurat_clusters %in% c("2","3"), "astrocyte", "glioma"))

DimPlot(scRNA.assays$MGH107, group.by = "prim_ident", label = T)
```





















```{r}
x = c(,,,)
GSE89567 = scRNA.assays
save(GSE89567, file = "GSE89567_SCT.Rdata")
```

```{r}
load(file = "GSE89567_SCT.Rdata")
```

```{r}
rm(list = ls()[!(ls() %in% c("f", "flip", "dropNA", "GSE89567"))])
```



```{r}
GSE89567 = GSE89567[!names(GSE89567) == "MGH57"]
GSE89567

x = SelectIntegrationFeatures(object.list = GSE89567, nfeatures = 3000)

scRNA = PrepSCTIntegration(object.list = GSE89567, anchor.features = x, verbose = T)

scRNA<- FindIntegrationAnchors(object.list = scRNA, normalization.method = "SCT", anchor.features = x, verbose = T)
scRNA <- IntegrateData(anchorset = scRNA, normalization.method = "SCT", verbose = T)
```


```{r}
dim = 50
scRNA <- RunPCA(scRNA, verbose = FALSE, npcs = dim)
dims = pcaMarchenkoPastur(M = length(VariableFeatures(scRNA)), N = dim(scRNA)[2], pca.sdev = scRNA@reductions$pca@stdev, factor = 1)
dims = (1:dim)[dims]
```


```{r}
dims = 1:30
scRNA <- RunUMAP(scRNA, dims = dims, verbose = FALSE)
scRNA <- FindNeighbors(scRNA, dims = dims, verbose = FALSE)
scRNA <- FindClusters(scRNA, verbose = FALSE)

el="integrated_snn_res.0.8"
DimPlot(scRNA,group.by = el, pt.size = 1, reduction = "umap", label = T) + scale_color_manual(values = alpha(rainbow(length(unique(dropNA(scRNA[[]][,el])))), 0.5), na.value = "gray")

#DimPlot(scRNA, group.by = "prim_ident")
```

```{r}
x = scRNA$prim_ident
x[scRNA$integrated_snn_res.0.8 %in% c("0","1", "2", "4", "5", "9")] = "glioma"
x[scRNA$integrated_snn_res.0.8 %in% c("8")] = "astrocyte"
x[scRNA$integrated_snn_res.0.8 %in% c("7")] = "astrocyte?"
x[scRNA$integrated_snn_res.0.8 %in% c("3", "6", "11")] = "immune"
x[scRNA$integrated_snn_res.0.8 %in% c("10")] = "oligodendroglia"
#x[scRNA$integrated_snn_res.0.8 %in% c("26")] = "neuron"
x[scRNA$integrated_snn_res.0.8 %in% c("12")] = "vascular?"
#x[scRNA$integrated_snn_res.0.8 %in% c("23")] = "OPC"
scRNA$identity = factor(x, levels = c("glioma", "astrocyte","astrocyte?", "immune", "oligodendroglia", "neuron", "vascular?"))

Idents(scRNA) = scRNA$identity

DimPlot(scRNA, group.by = "identity", reduction = "umap", pt.size = 2, label = F) + 
  scale_color_manual("Cell type", values = alpha(c("gray", hcl.colors(length(unique(scRNA$identity))-1, "Pastel 1")),0.5))
```

```{r}
x = scRNA[[]]
save(x, file = "GSE89567_meta.Rdata")
```

```{r}
xtabs(~prim_ident + identity, data = x)
```

```{r}
setwd("D:/Files/Main_directory/Nencki/Epilepsy_2020/dt")
#library(xlsx)
x = read.csv("markers_nature-2021.csv")

n = list()
for(el in colnames(x)){
  n[[el]] = x[,el][!x[,el] == ""]
}

cr = AUCell_buildRankings(scRNA@assays$RNA@counts)
AUC_n <- AUCell_calcAUC(x, cr)
scRNA[["AUC_n"]] = CreateAssayObject(counts = AUC_n@assays@data$AUC)

rownames(scRNA@assays$AUC_n@counts)
```

```{r}
f = function(x){
  x = (x-min(x))/(max(x) - min(x))
}


DefaultAssay(scRNA) = "AUC_n"
el = "Cahoy-JNeurosci-2011-astro-mature"
p = FeaturePlot(scRNA, features = el, pt.size = 1)
if(sum(scRNA@assays$AUC_n@counts==0)>0){
  p = p + scale_color_gradientn(colors = alpha(c("gray", hcl.colors(n = 5, "plasma")), 0.5),
                                values = f(c(0, quantile(scRNA@assays$AUC_n@counts[el,]))))
}else {
  p = p + scale_color_gradientn(colors = alpha(hcl.colors(n = 5, "plasma"),0.5), values = f(quantile(scRNA@assays$AUC_n@counts[el,])))
}
DefaultAssay(scRNA) = "integrated"

p
```

```{r}

x = FindMarkers(scRNA, group.by = "integrated_snn_res.0.8", ident.1 = "12", only.pos = T)
x
```



```{r}
#DimPlot(scRNA)
DefaultAssay(scRNA) = "RNA"
FeaturePlot(scRNA, "PTPRC")
DefaultAssay(scRNA) = "integrated"
```


```{r}
x = c(,,,)
dropNA = function(x){
  x = x[!is.na(x)]
  return(x)
}

MoMa = scRNA.assays
MoMa = lapply(MoMa, function(mscRNA){
  mscRNA[["Mo_Ma"]] = ifelse(mscRNA$singleR=="Monocytes"|mscRNA$singleR=="Macrophages", "yes", "no")
  #mscRNA = subset(mscRNA, subset = Mo_Ma == "yes")
  return(mscRNA)
})

MoMa = MoMa[sapply(MoMa, function(X){sum(dropNA(X$Mo_Ma) == "yes")})>0]

MoMa = lapply(MoMa, function(mscRNA){
  #mscRNA[["Mo_Ma"]] = ifelse(mscRNA$singleR=="Monocytes"|mscRNA$singleR=="Macrophages", "yes", "no")
  mscRNA = subset(mscRNA, subset = Mo_Ma == "yes")
  return(mscRNA)
})

save(MoMa, file = "MoMa_GSE89567.Rdata")
MoMa
```

```{r}
scRNA.assays

scRNA.assays = lapply(scRNA.assays, function(mscRNA){
  mscRNA <- subset(mscRNA, subset = Doublet == "no" & percent.mt < 10 & nFeature_RNA > 400)
  return(mscRNA)
})

scRNA.assays=scRNA.assays[sapply(scRNA.assays, function(x){dim(x@assays$RNA@meta.features)[1]})>10000]
scRNA.assays=scRNA.assays[sapply(scRNA.assays, function(x){length(x$orig.ident)})>200]

scRNA.assays
```


```{r}
for (i in 1:length(scRNA.assays)) {
  #DefaultAssay(scRNA.assays[[i]]) = "RNA"
  scRNA.assays[[i]] <- NormalizeData(scRNA.assays[[i]], verbose = FALSE)
  scRNA.assays[[i]] <- FindVariableFeatures(scRNA.assays[[i]], selection.method = "vst", nfeatures = 2000, verbose = FALSE)
}
save(scRNA.assays, file = "scRNAassays3.Rdata")

memory.limit(25*16257)
options(future.globals.maxSize = 8*500*1024^2)
scRNA <- FindIntegrationAnchors(object.list = scRNA.assays, dims = 1:30)

#save(scRNA, file="anchors.Rdata")

scRNA <- IntegrateData(anchorset = scRNA, dims = 1:30)
save(scRNA, file = "scRNAintegrated1.Rdata")

dim = 125
scRNA <- ScaleData(scRNA)
scRNA <- FindVariableFeatures(scRNA)
scRNA <- RunPCA(scRNA, features = VariableFeatures(object = scRNA), npcs = dim)
'scRNA <- RunUMAP(scRNA, dims = 1:10)
scRNA <- RunTSNE(scRNA, dims = 1:10)'

#setwd(paste(wd, "//merge", sep = ""))
#scRNA <- RunPCA(scRNA, features = VariableFeatures(object = scRNA), npcs = dim)
scRNA <- JackStraw(scRNA, dims = dim, num.replicate = 125)

scRNA <- ScoreJackStraw(scRNA, dims = 1:dim, score.thresh = 1e-10, do.plot = F)

p = JackStrawPlot(scRNA, dims = 1:dim) + scale_color_manual(values = rainbow(dim)[rep(dim*c(0, 0.2, 0.4, 0.6, 0.8), times = 20)+ unlist(lapply(1:(dim/5), rep, times = 5))])
p

ggsave(paste0("JackStrawPlot", dim, "150.png"), p, width = 32, height = 6, dpi = 400)
ggsave(paste0("JackStrawPlot", dim, "150.pdf"), p, width = 32, height = 6, dpi = 400)

dims = (1:dim)[as.numeric(sapply(strsplit(as.character(unique(ggplot_build(p)$plot$data$PC.Score)), ": "), function(X){return(X[2])}))<0.05]

scRNA <- RunUMAP(scRNA, dims = dims)
scRNA <- RunTSNE(scRNA, dims = dims)

library(ggplot2)
ggsave(plot = DimPlot(scRNA, reduction = "umap", label = F),
       filename = "Samples after integration.pdf",
       width = 8, height = 6, dpi = 400)

ggsave(plot = DimPlot(scRNA, reduction = "umap", label = F),
       filename = "Samples after integration.png",
       width = 8, height = 6, dpi = 400)

scRNA <- FindNeighbors(scRNA, dims = dims)
for(el in 0.1 + 0:14/10){
  scRNA <- FindClusters(scRNA, resolution = el)
  'if(sum(grepl(paste("integrated_snn_res.", el, sep = ""), colnames(scRNA[[]])))==0){
    scRNA <- FindClusters(scRNA, resolution = el)
  }'
}

save(scRNA, file = "scRNA_clusters.Rdata")

library(clustree)

p = clustree(scRNA, prefix = "integrated_snn_res.")

ggsave(plot = p, filename = paste("Clustree.png",sep = ""), width = 10, height = 8, dpi = 400)
ggsave(plot = p, filename = paste("Clustree.pdf",sep = ""), width = 10, height = 8, dpi = 400)
print(dims)
```


```{r}
```

```{r}
load("scRNA_clusters.Rdata")
scRNA <- BuildClusterTree(scRNA)

PlotClusterTree(scRNA)
```

```{r}
scRNA <- AddModuleScore(scRNA, features =  list(mdsc = c("S100A8", "S100A9", "CEBPB", "ARG1", "ARG2", "IL1B", "IDO1", "NOS2", "IL4R", "IL10", "TGFB1")), name = "MDSC", ctrl = 75, assay = "RNA")
```
```{r}
p = DimPlot(scRNA, reduction = "umap", label = F, group.by = "singleR", pt.size = 2)
x = setNames(rep("", length(levels(ggplot_build(p)$plot$data$singleR))), levels(ggplot_build(p)$plot$data$singleR))
x[!(levels(ggplot_build(p)$plot$data$singleR)=="Monocytes"|levels(ggplot_build(p)$plot$data$singleR)=="Macrophages")] = rainbow(length(levels(ggplot_build(p)$plot$data$singleR))-2)
x["Macrophages"] = "black"
x["Monocytes"] = "gray30"
p = p + scale_color_manual(values=alpha(x, 0.5)) + ggtitle("GSE89567 - singleR")


p = grid.arrange(p, FeaturePlot(scRNA, features = "MDSC1", pt.size = 2) + ggtitle("GSE89567 - MDSC score") + scale_color_gradientn(colors = alpha(brewer.pal(n = 6, name = "YlGnBu"),0.5)), ncol = 2)

ggsave(p, filename = "GSE89567 before selection.png", dpi = 400, width = 16, height = 6)
```
```{r}
p = DimPlot(scRNA, reduction = "tsne", label = F, group.by = "singleR", pt.size = 2)
x = setNames(rep("", length(levels(ggplot_build(p)$plot$data$singleR))), levels(ggplot_build(p)$plot$data$singleR))
x[!(levels(ggplot_build(p)$plot$data$singleR)=="Monocytes"|levels(ggplot_build(p)$plot$data$singleR)=="Macrophages")] = rainbow(length(levels(ggplot_build(p)$plot$data$singleR))-2)
x["Macrophages"] = "black"
x["Monocytes"] = "gray30"
p = p + scale_color_manual(values=alpha(x, 0.5)) + ggtitle("GSE89567 - singleR")


p = grid.arrange(p, FeaturePlot(scRNA, features = "MDSC1", pt.size = 2, reduction = "tsne") + ggtitle("GSE89567 - MDSC score") + scale_color_gradientn(colors = alpha(brewer.pal(n = 6, name = "YlGnBu"),0.5)), ncol = 2)

ggsave(p, filename = "GSE89567 before selection tSNE.png", dpi = 400, width = 16, height = 6)
```

```{r}
VlnPlot(scRNA, features = "MDSC1", group.by = "integrated_snn_res.0.1")
```


```{r}
ct = scRNA@reductions$umap@cell.embeddings
ct = cbind(ct, scRNA[[]][rownames(ct),grepl("integrated_snn_res",colnames(scRNA[[]]))])

for(el in colnames(ct)[grepl("integrated_snn_res",colnames(ct))]){
  ct[,el] = as.numeric(as.character(ct[,el]))
}
colnames(ct) = gsub("integrated_snn_res.", "K", colnames(ct))

x = as.numeric(gsub("K", "", colnames(ct)))
x[is.na(x)] = 0
x = 10*x

ct = ct[,(!grepl("K", colnames(ct)))|(x%%2==1)]

p = clustree_overlay(ct, prefix = "K", x_value = "UMAP_1", y_value = "UMAP_2", label_nodes = T)
p = p + theme_bw()
ggsave(p, width = 12, height = 8, dpi = 400, filename = "clustree_UMAP.png")
```


```{r}
p = DimPlot(scRNA, reduction = "umap", label = F, group.by = "singleR")
x = terrain.colors(length(levels(ggplot_build(p)$plot$data$singleR)))
x[levels(ggplot_build(p)$plot$data$singleR)=="Macrophages"] = "cornflowerblue"
x[levels(ggplot_build(p)$plot$data$singleR)=="Monocytes"] = "blue4"
p + scale_color_manual(values=x)
```

```{r}
xtabs(~singleR+integrated_snn_res.0.1, data = scRNA[[]])
```


```{r}
p = DimPlot(scRNA, reduction = "umap", label = F, group.by = "orig.ident")
x = rainbow(length(levels(ggplot_build(p)$plot$data$orig.ident)))
p + scale_color_manual(values=x)
```

```{r}
DimPlot(scRNA, reduction = "umap", label = T)
scRNA[["sel"]] = ifelse(scRNA[[]]$integrated_snn_res.0.1==1, "yes", "no")
#scRNA[["sel"]] = ifelse(scRNA[[]]$singleR=="Macrophages"|scRNA[[]]$singleR=="Monocytes", "yes", "no")
#table(scRNA[[]][scRNA[[]]$sel=="yes","Cell_type"])
table(scRNA[[]][scRNA[[]]$sel=="yes","singleR"])
```
```{r}
wd = getwd()
setwd(paste(wd, "//MoMa", sep = ""))
scRNA = subset(scRNA, subset = sel == "yes")
save(scRNA, file = "scRNA1.Rdata")
```



```{r}
load("GSE84465_scRNA_clusters.Rdata")
```


```{r}
wd = getwd()
setwd(paste(wd, "//MoMa", sep = ""))
scRNA = subset(scRNA, subset = sel == "yes")
save(scRNA, file = "scRNA1.Rdata")
```

```{r}
dim = 125
scRNA <- ScaleData(scRNA)
scRNA <- FindVariableFeatures(scRNA)
scRNA <- RunPCA(scRNA, features = VariableFeatures(object = scRNA), npcs = dim)
'scRNA <- RunUMAP(scRNA, dims = 1:10)
scRNA <- RunTSNE(scRNA, dims = 1:10)'

#setwd(paste(wd, "//merge", sep = ""))
#scRNA <- RunPCA(scRNA, features = VariableFeatures(object = scRNA), npcs = dim)
scRNA <- JackStraw(scRNA, dims = dim, num.replicate = 150)

scRNA <- ScoreJackStraw(scRNA, dims = 1:dim, score.thresh = 1e-10, do.plot = F)

p = JackStrawPlot(scRNA, dims = 1:dim) + scale_color_manual(values = rainbow(dim)[rep(dim*c(0, 0.2, 0.4, 0.6, 0.8), times = 20)+ unlist(lapply(1:(dim/5), rep, times = 5))])
p

ggsave(paste0("JackStrawPlot", dim, ".png"), p, width = 32, height = 6, dpi = 400)
ggsave(paste0("JackStrawPlot", dim, ".pdf"), p, width = 32, height = 6, dpi = 400)

dims = (1:dim)[as.numeric(sapply(strsplit(as.character(unique(ggplot_build(p)$plot$data$PC.Score)), ": "), function(X){return(X[2])}))<0.05]

scRNA <- RunUMAP(scRNA, dims = dims)
scRNA <- RunTSNE(scRNA, dims = dims)

setwd(paste(wd, "//MoMa", sep = ""))
library(ggplot2)
ggsave(plot = DimPlot(scRNA, reduction = "umap", label = F),
       filename = "Samples after integration.pdf",
       width = 8, height = 6, dpi = 400)

ggsave(plot = DimPlot(scRNA, reduction = "umap", label = F),
       filename = "Samples after integration.png",
       width = 8, height = 6, dpi = 400)

scRNA <- FindNeighbors(scRNA, dims = dims)
for(el in 0.1 + 0:14/10){
  scRNA <- FindClusters(scRNA, resolution = el)
  'if(sum(grepl(paste("integrated_snn_res.", el, sep = ""), colnames(scRNA[[]])))==0){
    scRNA <- FindClusters(scRNA, resolution = el)
  }'
}

save(scRNA, file = "scRNAMoMa_clusters.Rdata")

library(clustree)

p = clustree(scRNA, prefix = "integrated_snn_res.")

ggsave(plot = p, filename = paste("Clustree.png",sep = ""), width = 10, height = 8, dpi = 400)
ggsave(plot = p, filename = paste("Clustree.pdf",sep = ""), width = 10, height = 8, dpi = 400)
print(dims)
```


```{r}
ct = scRNA@reductions$umap@cell.embeddings
ct = cbind(ct, scRNA[[]][rownames(ct),grepl("integrated_snn_res",colnames(scRNA[[]]))])

for(el in colnames(ct)[grepl("integrated_snn_res",colnames(ct))]){
  ct[,el] = as.numeric(as.character(ct[,el]))
}
colnames(ct) = gsub("integrated_snn_res.", "K", colnames(ct))

x = as.numeric(gsub("K", "", colnames(ct)))
x[is.na(x)] = 0
x = 10*x

ct = ct[,(!grepl("K", colnames(ct)))|(x%%2==1)]

p = clustree_overlay(ct, prefix = "K", x_value = "UMAP_1", y_value = "UMAP_2", label_nodes = T)
p = p + theme_bw()
setwd(paste(wd, "//MoMa", sep = ""))
ggsave(p, width = 12, height = 8, dpi = 400, filename = "clustree_UMAP.png")
```

```{r}
x = sapply(unique(scRNA$orig.ident), function(X){mean(scRNA@reductions$umap@cell.embeddings[,"UMAP_1"][scRNA$orig.ident==X])})
x = x[order(x)]

X = as.factor(as.character(scRNA$orig.ident))

for(el in names(x)){
  X = relevel(X, el)
}

scRNA$orig.ident = X

DimPlot(scRNA, group.by = "orig.ident", cols = "Reds") + theme(panel.background = element_rect(fill = "gray"))
```

```{r}
DimPlot(scRNA, group.by = "integrated_snn_res.0.8", cols = "Set2")
```

```{r}
DimPlot(scRNA, reduction = "tsne", cols = "Set2")
```

```{r}
wd = getwd()
setwd(paste(wd, "//MoMa", sep = ""))
load(file = "scRNAMoMa_clusters.Rdata")
scRNA@assays$RNA@counts[1:10,1:10]
```

```{r}
Idents(scRNA) = scRNA$integrated_snn_res.0.8

x = sapply(levels(Idents(scRNA)), function(el){
  mean(scRNA@assays$RNA@counts["TMEM119",rownames(scRNA[[]][Idents(scRNA)==el,])])
})

x = x[order(x)]

#X = factor(paste("C", Idents(scRNA), sep = ""))
X = Idents(scRNA)

for(el in names(x)){
  X = relevel(X, el)
}  

library(plyr)
X = mapvalues(X, to = 0:max(as.numeric(as.character(levels(X)))), from = levels(X))

Idents(scRNA) = X

DimPlot(scRNA, reduction = "umap", label = T)

scRNA.markers = FindAllMarkers(scRNA)
```

```{r}
p = DimPlot(scRNA, reduction = "umap", label = F, group.by = "singleR")
x = terrain.colors(length(levels(ggplot_build(p)$plot$data$singleR)))
x[levels(ggplot_build(p)$plot$data$singleR)=="Macrophages"] = "cornflowerblue"
x[levels(ggplot_build(p)$plot$data$singleR)=="Monocytes"] = "blue4"
p + scale_color_manual(values=x)
```

```{r}
Panglao = read.table(file = "PanglaoDB_markers_23_Dec_2019.tsv", h=T, sep = '\t', quote = '')

Panglao = Panglao[grepl("Hs", Panglao$species),]
PG = list()
for(el in unique(Panglao$cell.type)){
  PG[[el]] = as.character(Panglao[Panglao$cell.type==el,"official.gene.symbol"])
}
library(xlsx)
markers = read.xlsx(file = "Gene_markers.xlsx", sheetName = "Sheet1")
markers$gene = toupper(markers$gene)



wd = getwd()
setwd(paste(wd, "//MoMa", sep = ""))
markers = markers[markers$ofi=="yes",]
el = as.character(markers$group)
el = sapply(el, function(x){
  paste(toupper(substr(x,1,1)), substr(x,2,nchar(x)), sep = "")
})

el = sapply(el, function(x){
  if(nchar(x)>25){
    x = paste(substr(x, 1, which(strsplit(x, "")[[1]]==" ")[which.min(abs((2*which(strsplit(x, "")[[1]]==" ")/nchar(x))-1))]-1),substr(x, which(strsplit(x, "")[[1]]==" ")[which.min(abs((2*which(strsplit(x, "")[[1]]==" ")/nchar(x))-1))]+1, nchar(x)), sep = "\n")
  }
  return(x)
})
markers$group = as.factor(el)

p = DotPlot(scRNA, features = unique(as.character(markers$gene)), assay = "RNA", cols = "RdYlBu")
p = p + RotatedAxis()
p$data$group = as.character(markers$group)[match(as.character(p$data$features.plot), as.character(markers$gene))]

p$data$group = factor(p$data$group, order = F)

for(el in rev(unique(as.character(markers$group)))){
  if(sum(p$data$group==el)>0){
    p$data$group = relevel(p$data$group, ref = el)
  }
}

p = p + facet_grid(. ~ group, drop = T, scales = "free_x", space = "free_x") +theme(strip.text.x = element_text(angle=90))
x = length(unique(ggplot_build(p)$plot$data$features.plot))/5
p1 = DimPlot(scRNA, reduction = "umap", label = T, cols = "Set2") + ggtitle("Macrophages - assigned clusters")
library(gridExtra)
p = grid.arrange(p1,p, nrow = 1, widths = c(9, 3+x))

ggsave(p, filename = "markers.pdf", width = 12+x, height = 7)
ggsave(p, filename = "markers.png", width = 12+x, height = 7)


panglao = PG
panglao = panglao[c("Myeloid-derived suppressor cells", "Macrophages", "Monocytes", "Microglia", "Neutrophils")]
panglao = sapply(panglao, function(x){x = intersect(x, rownames(scRNA))})
panglao = data.frame(gene = unlist(panglao), group = rep(names(panglao), times = sapply(panglao, length)))
p = DotPlot(scRNA, features = unique(as.character(panglao$gene)), assay = "RNA", cols = "RdYlBu")
p = p + RotatedAxis()
p$data$group = as.character(panglao$group)[match(as.character(p$data$features.plot), as.character(panglao$gene))]

p$data$group = factor(p$data$group, order = F)

for(el in rev(unique(as.character(panglao$group)))){
  if(sum(p$data$group==el)>0){
    p$data$group = relevel(p$data$group, ref = el)
  }
}

p = p + facet_grid(. ~ group, drop = T, scales = "free_x", space = "free_x") +theme(strip.text.x = element_text(angle=90))
x = length(unique(ggplot_build(p)$plot$data$features.plot))/5
p1 = DimPlot(scRNA, reduction = "umap", label = T, cols = "Set2") + ggtitle("Macrophages - assigned clusters")

p = grid.arrange(p1,p, nrow = 1,  widths = c(9, 3+x))

ggsave(p, filename = "panglao.pdf", width = 12+x, height = 7)
ggsave(p, filename = "panglao.png", width = 12+x, height = 7)
```

```{r}
x = scRNA[[]]
save(x, file = "GSE89567_clusters.Rdata")
```

```{r}
intersect(markers[grepl("mmdsc", tolower(markers$group)),"gene"], scRNA.markers[scRNA.markers$p_val_adj<0.05,]$gene)
intersect(markers[grepl("pmnmdsc", tolower(markers$group)),"gene"], scRNA.markers[scRNA.markers$p_val_adj<0.05,]$gene)
intersect(markers[grepl("microglia", tolower(markers$group)),"gene"], scRNA.markers[scRNA.markers$p_val_adj<0.05,]$gene)
intersect(markers[grepl("border", tolower(markers$group)),"gene"], scRNA.markers[scRNA.markers$p_val_adj<0.05,]$gene)
```

```{r}
DoHeatmap(scRNA,
          features = unique(c(intersect(markers[grepl("mmdsc", tolower(markers$group)),"gene"], scRNA.markers[scRNA.markers$p_val_adj<0.05,]$gene),
                              intersect(markers[grepl("pmnmdsc", tolower(markers$group)),"gene"], scRNA.markers[scRNA.markers$p_val_adj<0.05,]$gene),
                              intersect(markers[grepl("microglia", tolower(markers$group)),"gene"], scRNA.markers[scRNA.markers$p_val_adj<0.05,]$gene),
                              intersect(markers[grepl("border", tolower(markers$group)),"gene"], scRNA.markers[scRNA.markers$p_val_adj<0.05,]$gene))))
```


```{r}
scRNA.markers[scRNA.markers$gene %in% markers$gene,]
```




```{r}
wd = getwd()
setwd(paste(wd, "//MoMa", sep = ""))

library(gridExtra)
p1 = list()
for(el in c("HLA-DRA", "CD14", "CD33", "S100A8", "S100A9", "CD274", "IL1B", "IL10", "CD84")){
  p = FeaturePlot(scRNA, features = el, slot = "counts")
  x = ggplot_build(p)$plot$data[,4]
  p = p + scale_color_gradientn(colours = c("gray", "cornflowerblue", "blue4", "red4", "red"),values = (c(min(x),
      min(x[!x==0]),
      median(x[!x==0]),
      mean(x[x>(median(x[!x==0]))]),
      max(x[!x==0]))-min(x))/max(x[!x==0]))
  p1[[el]] = p
}


ggsave(do.call("grid.arrange", c(p1, ncol=3)), filename = "MMDSC_markers.png", width = 24, height = 18, dpi = 400)

p1 = list()
for(el in c("CD14","S100A8", "S100A9", "OLR1",   "XBP1",   "IL1B",   "CD84", "TGFB1")){
  p = FeaturePlot(scRNA,  features = el, slot = "counts")
  x = ggplot_build(p)$plot$data[,4]
  p = p + scale_color_gradientn(colours = c("gray", "cornflowerblue", "blue4", "red4", "red"),values = (c(min(x),
      min(x[!x==0]),
      median(x[!x==0]),
      mean(x[x>(median(x[!x==0]))]),
      max(x[!x==0]))-min(x))/max(x[!x==0]))
  p1[[el]] = p
}

setwd(paste(wd, "//MoMa", sep = ""))
ggsave(do.call("grid.arrange", c(p1, ncol=4)), filename = "PMNMDSC_markers.png", width = 32, height = 12, dpi = 400)

p1 = list()
for(el in c("CCL3", "TMEM119", "CX3CR1", "P2RY12", "P2RY13", "GPR34", "OLFML3", "SELPLG", "SPARC", "SLC2A5", "F13A1", "IFIT3", "CSF1", "CD14", "LPL", "CST7", "TGM2", "GPNMB")){
  p = FeaturePlot(scRNA,  features = el, slot = "counts")
  x = ggplot_build(p)$plot$data[,4]
  p = p + scale_color_gradientn(colours = c("gray", "cornflowerblue", "blue4", "red4", "red"),values = (c(min(x),
      min(x[!x==0]),
      median(x[!x==0]),
      mean(x[x>(median(x[!x==0]))]),
      max(x[!x==0]))-min(x))/max(x[!x==0]))
  p1[[el]] = p
}

setwd(paste(wd, "//MoMa", sep = ""))
ggsave(do.call("grid.arrange", c(p1, ncol=3)), filename = "Microglia_markers.png", width = 24, height = 36, dpi = 400)

p1 = list()
for(el in c("APOE", "CSF1R", "CD163", "CX3CR1", "LYVE1", "MS4A7")){
  p = FeaturePlot(scRNA,  features = el, slot = "counts")
  x = ggplot_build(p)$plot$data[,4]
  p = p + scale_color_gradientn(colours = c("gray", "cornflowerblue", "blue4", "red4", "red"),values = (c(min(x),
      min(x[!x==0]),
      median(x[!x==0]),
      mean(x[x>(median(x[!x==0]))]),
      max(x[!x==0]))-min(x))/max(x[!x==0]))
  p1[[el]] = p
}

setwd(paste(wd, "//MoMa", sep = ""))
ggsave(do.call("grid.arrange", c(p1, ncol=3)), filename = "BAM_markers.png", width = 24, height = 12, dpi = 400)
```


```{r}
DoHeatmap(scRNA, features = (intersect(PG$Macrophages, scRNA.markers[scRNA.markers$p_val_adj<0.05,]$gene)))
```

```{r}
DoHeatmap(scRNA, features = (intersect(PG$Monocytes, scRNA.markers[scRNA.markers$p_val_adj<0.05,]$gene)))
```

```{r}
DoHeatmap(scRNA, features = (intersect(PG$`Myeloid-derived suppressor cells`, scRNA.markers[scRNA.markers$p_val_adj<0.05,]$gene)))
```

```{r}
DoHeatmap(scRNA, features = (intersect(PG$Microglia, scRNA.markers[scRNA.markers$p_val_adj<0.05,]$gene)))
```

```{r}
library(TSCAN)
preprocess <- preprocess(as.matrix(scRNA@assays$RNA@counts), minexpr_percent = 0.4,  cvcutoff = 0.6)
dim(preprocess)
```
```{r}
tscan <- exprmclust(preprocess)

scRNA[["tscan"]] = tscan$clusterid[rownames(scRNA[[]])]

DimPlot(scRNA, group.by = "tscan")
```
```{r}
plotmclust(tscan, show_cell_names = F)
```
```{r}
diffval <- difftest(preprocess,TSCANorder(tscan))

diffval[order(diffval$qval),]
```

```{r}
intersect(rownames(diffval[diffval$qval<0.05,]), markers$gene)
```
```{r}
expr <- log2(as.matrix(scRNA@assays$RNA@counts)["XBP1",]+1)
singlegeneplot(expr, TSCANorder(tscan,flip=TRUE,orderonly=FALSE))
```

```{r}
  p = FeaturePlot(scRNA,  features = "TMEM119")
  x = ggplot_build(p)$plot$data[,4]
  p = p + scale_color_gradientn(colours = c("cornflowerblue", "blue4", "red4", "red", "yellow"))
  p
```
```{r}
scRNA[["tscan_pca"]] <- CreateDimReducObject(embeddings = tscan$pcareduceres, key = "TSCAN_", assay = DefaultAssay(scRNA))
```

```{r}
wd = getwd()
setwd(paste(wd, "//MoMa", sep = ""))
Idents(scRNA) = scRNA[[]]$tscan
markers = markers[markers$ofi=="yes",]
el = as.character(markers$group)
el = sapply(el, function(x){
  paste(toupper(substr(x,1,1)), substr(x,2,nchar(x)), sep = "")
})

el = sapply(el, function(x){
  if(nchar(x)>25){
    x = paste(substr(x, 1, which(strsplit(x, "")[[1]]==" ")[which.min(abs((2*which(strsplit(x, "")[[1]]==" ")/nchar(x))-1))]-1),substr(x, which(strsplit(x, "")[[1]]==" ")[which.min(abs((2*which(strsplit(x, "")[[1]]==" ")/nchar(x))-1))]+1, nchar(x)), sep = "\n")
  }
  return(x)
})
markers$group = as.factor(el)

p = DotPlot(scRNA, features = unique(as.character(markers$gene)), assay = "RNA", cols = "RdYlBu")
p = p + RotatedAxis()
p$data$group = as.character(markers$group)[match(as.character(p$data$features.plot), as.character(markers$gene))]

p$data$group = factor(p$data$group, order = F)

for(el in rev(unique(as.character(markers$group)))){
  if(sum(p$data$group==el)>0){
    p$data$group = relevel(p$data$group, ref = el)
  }
}

p = p + facet_grid(. ~ group, drop = T, scales = "free_x", space = "free_x") +theme(strip.text.x = element_text(angle=90))
x = length(unique(ggplot_build(p)$plot$data$features.plot))/5
p1 = DimPlot(scRNA, reduction = "tscan_pca", label = T, cols = "Set2") + ggtitle("Macrophages - TSCAN groups")

p = grid.arrange(p1,p, nrow = 1, widths = c(9, 3+x))

ggsave(p, filename = "markers_TSCAN.pdf", width = 12+x, height = 7)
ggsave(p, filename = "markers_TSCAN.png", width = 12+x, height = 7)


panglao = PG
panglao = panglao[c("Myeloid-derived suppressor cells", "Macrophages", "Monocytes", "Microglia", "Neutrophils")]
panglao = sapply(panglao, function(x){x = intersect(x, rownames(scRNA))})
panglao = data.frame(gene = unlist(panglao), group = rep(names(panglao), times = sapply(panglao, length)))
p = DotPlot(scRNA, features = unique(as.character(panglao$gene)), assay = "RNA", cols = "RdYlBu")
p = p + RotatedAxis()
p$data$group = as.character(panglao$group)[match(as.character(p$data$features.plot), as.character(panglao$gene))]

p$data$group = factor(p$data$group, order = F)

for(el in rev(unique(as.character(panglao$group)))){
  if(sum(p$data$group==el)>0){
    p$data$group = relevel(p$data$group, ref = el)
  }
}

p = p + facet_grid(. ~ group, drop = T, scales = "free_x", space = "free_x") +theme(strip.text.x = element_text(angle=90))
p1 = DimPlot(scRNA, reduction = "tscan_pca", label = T, cols = "Set2") + ggtitle("Macrophages - TSCAN groups")

p = grid.arrange(p1,p, nrow = 1,  widths = c(9, 3+x))

ggsave(p, filename = "panglao_TSCAN.pdf", width = 12+x, height = 7)
ggsave(p, filename = "panglao_TSCAN.png", width = 12+x, height = 7)
```

```{r}
scRNA@assays$integrated@scale.data
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
