---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

```{r}
rm(list=ls())
```

```{r}
dropNA = function(x){
  x = x[!is.na(x)]
  return(x)
}

flip <- function(x){
  x = as.matrix(x)
  y = t(apply(x, 2, rev))
  rownames(y)=colnames(x)
  colnames(y)=rev(rownames(x))
  return (y)
}

gexp = function(scRNA = scRNA, gene, reduction = "UMAP"){
  if(reduction %in% c("TSNE", "UMAP", "PCA")){
    reduction = tolower(reduction)
  }
  
  df = data.frame(scale = scRNA@assays$integrated@scale.data[gene,rownames(scRNA[[]])],
                  counts = scRNA@assays$RNA@counts[gene,rownames(scRNA[[]])],
                  x = scRNA@reductions[[reduction]]@cell.embeddings[rownames(scRNA[[]]),1],
                  y = scRNA@reductions[[reduction]]@cell.embeddings[rownames(scRNA[[]]),2])
  
  df$value = df$scale
  df$value = df$value + 0.01 - min(df$value[df$counts>0])
  df$value[df$counts==0] = 0
  p = ggplot(df, aes(x = x, y = y, color = value))+
    geom_point() + labs(title = gene, x = paste(reduction, 1, sep = "_"), y = paste(reduction, 2, sep = "_")) +
    scale_color_gradientn(name = "Scaled\nexpression\nlevel",
                          values = c(0, quantile(df$value[df$value>0]))/max(df$value),
                          colors = alpha(c("gray", rev(hcl.colors(5, "RdYlBu"))),0.5))+
    theme_classic()
  return(p)
}
```


```{r}
library(GEOquery)

GSE84465 = getGEO("GSE84465")

#filePaths = getGEOSuppFiles("GSE84465")
```

```{r}
library(Seurat)
library(xlsx)
library(Matrix)
library(future)
library(biomaRt)
library(ggplot2)
library(scds)
library(SingleR)
#plan("multiprocess", workers = 4)
```

```{r}
mart <- useMart("ENSEMBL_MART_ENSEMBL")
ensembl = useDataset("hsapiens_gene_ensembl",mart=mart)
trl = getBM(attributes=c("ensembl_gene_id","hgnc_symbol", "entrezgene_id", "gene_biotype", "start_position", "end_position", "chromosome_name", "strand", "description","phenotype_description", "mim_morbid_description"), mart= ensembl)
trl$chromosome_name[nchar(trl$chromosome_name)>3] = gsub("_", "", substr(trl$chromosome_name[nchar(trl$chromosome_name)>3], 10, 11))
```


```{r}

setwd(paste(getwd(), "/GSE84465", sep = ""))
meta = as.data.frame(GSE84465$GSE84465_series_matrix.txt.gz)
x = rownames(meta)

X = matrix(sapply(strsplit(as.character(read.csv(strsplit(as.character(meta[x[1],"supplementary_file_1"]),"/")[[1]][9])[,1]), " "), function(X){X[2]}), ncol = 1, dimnames = list(r = sapply(strsplit(as.character(read.csv(strsplit(as.character(meta[x[1],"supplementary_file_1"]),"/")[[1]][9], sep = "\t")[,1]), " "), function(X){X[1]}), c = x[1]))

for(el in x[-1]){
  X = cbind(X,
            (matrix(sapply(strsplit(as.character(read.csv(strsplit(as.character(meta[el,"supplementary_file_1"]),"/")[[1]][9])[,1]), " "), function(X){X[2]}), ncol = 1, dimnames = list(r = sapply(strsplit(as.character(read.csv(strsplit(as.character(meta[el,"supplementary_file_1"]),"/")[[1]][9], sep = "\t")[,1]), " "), function(X){X[1]}), c = el)))[rownames(X),])
}

#save(X, file = "X.Rdata")


```


```{r}
meta = as.data.frame(GSE84465$GSE84465_series_matrix.txt.gz)
meta
```



```{r}
x = rownames(X)
X = apply(X,2,as.numeric)
rownames(X) = x

X = X[rowSums(X)>0,]
```

```{r}
X = X[order(apply(X, 1, var), decreasing = T),]

library(HGNChelper)
x = checkGeneSymbols(rownames(X))

x$ss = sapply(strsplit(x$Suggested.Symbol, " /// "), function(X){X[1]})

#x$ss[is.na(x$ss)] = x[is.na(x$ss),"x"]

X = X[!is.na(x$Suggested.Symbol),]
x = x[!is.na(x$Suggested.Symbol),]

X = X[!duplicated(x$ss),]
x = x[!duplicated(x$ss),]

rownames(X) = x$ss
```


```{r}
setwd(paste(getwd(), "/GSE84465", sep = ""))
reference <- BlueprintEncodeData()
scRNA.ds = unique(meta$patient.id.ch1)
scRNA.assays = list()
for(el in scRNA.ds){
  x = X[,rownames(meta[meta$patient.id.ch1==el,])]
  
  x = CreateSeuratObject(counts = x, project = el, min.cells = 5, min.features = 200)
  x <- PercentageFeatureSet(x, pattern = "^MT-", col.name = "percent.mt")
  doublet = cxds_bcds_hybrid(as.SingleCellExperiment(x), estNdbl = T)
  common <- intersect(rownames(x@assays$RNA@data), rownames(reference))
  singler <- reference[common,]
  singler <- SingleR(test = as.matrix(x@assays$RNA@data[common,]), ref = singler, labels = singler$label.main)
  
  x[["singleR"]] = singler$pruned.labels
  x[["cxds_score"]] = doublet$cxds_score
  x[["bcds_score"]] = doublet$bcds_score
  x[["hybrid_score"]] = doublet$hybrid_score
  x[["Doublet"]] = ifelse(doublet$hybrid_call, "yes", "no")
  #x[["Histology"]] = gsub("genotype/variation: ", "", meta[match(el, meta$name),"X.Sample_characteristics_ch1.2"])
  x[["Age"]] = age[el]
  x[["IDH"]] = idh[el]
  x[["Tumor"]] = gsub("tissue: ", "", meta[colnames(x),"characteristics_ch1.3"])
  x[["Cell_type"]] = gsub("cell type: ", "", meta[colnames(x),"characteristics_ch1.6"])
  x[["Selection"]] = gsub("	selection: ", "", meta[colnames(x),"characteristics_ch1.8"])
  #assign(el, x) 
  
  scRNA.assays[[el]] = x   
}
#save(scRNA.assays, file = "scRNA.assays.Rdata")
```


```{r}
#Preparation
```

```{r}
library(Matrix)
library(Seurat)
load("scRNA.assays.Rdata")
#scRNA.assays

'for(el in names(scRNA.assays)){
  x = scRNA.assays[[el]]@assays$RNA@counts
  x = RelativeCounts(x, scale.factor = 1e6, verbose = TRUE)
  scRNA.assays[[el]]@assays$RNA@counts = x
  scRNA.assays[[el]]@assays$RNA@data = x
  print(el)
}'

scRNA.assays = lapply(scRNA.assays, function(mscRNA){
  mscRNA <- subset(mscRNA, subset = Doublet == "no" & percent.mt < 10 & nFeature_RNA > 400)
  return(mscRNA)
})

scRNA.assays=scRNA.assays[sapply(scRNA.assays, function(x){dim(x@assays$RNA@meta.features)[1]})>10000]
scRNA.assays=scRNA.assays[sapply(scRNA.assays, function(x){length(x$orig.ident)})>200]

ig = c("ACTA2", "APOLD1", "CD24", "PTPRC", "CLU", "MBP", "GPR17", "SNAP25")
for (i in 1:length(scRNA.assays)) {
  scRNA.assays[[i]] <- NormalizeData(scRNA.assays[[i]])
  scRNA.assays[[i]] <- FindVariableFeatures(scRNA.assays[[i]], selection.method = "vst", nfeatures = 3000)
  scRNA.assays[[i]] <- ScaleData(scRNA.assays[[i]], features = unique(c(ig, VariableFeatures(scRNA.assays[[i]]))))
  
  print(i)
}

library(URD)
D = list()
#p = list()
dim = 50
for (i in 1:length(scRNA.assays)) {
  scRNA.assays[[i]] <- RunPCA(scRNA.assays[[i]], verbose = FALSE, npcs = dim, features = VariableFeatures(scRNA.assays[[i]]))
  
  dims = pcaMarchenkoPastur(M = length(VariableFeatures(scRNA.assays[[i]])), N = dim(scRNA.assays[[i]])[2], pca.sdev = scRNA.assays[[i]]@reductions$pca@stdev, factor = 1)
  dims = (1:dim)[dims]
  
  scRNA.assays[[i]] <- RunUMAP(scRNA.assays[[i]], dims = dims, verbose = FALSE)
  
  scRNA.assays[[i]] <- FindNeighbors(scRNA.assays[[i]], dims = dims, verbose = FALSE)
  scRNA.assays[[i]] <- FindClusters(scRNA.assays[[i]], verbose = FALSE)
  
  D[[i]] = dims
  #p[[i]] = DimPlot(scRNA.assays[[i]], label = TRUE) + NoLegend()
}

gexp = function(scRNA = scRNA, gene, reduction = "UMAP"){
  df = data.frame(scale = scRNA@assays$RNA@scale.data[gene,rownames(scRNA[[]])],
                  counts = scRNA@assays$RNA@counts[gene,rownames(scRNA[[]])],
                  x = scRNA@reductions[[tolower(reduction)]]@cell.embeddings[rownames(scRNA[[]]),1],
                  y = scRNA@reductions[[tolower(reduction)]]@cell.embeddings[rownames(scRNA[[]]),2])
  
  df$value = df$scale
  df$value = df$value + 0.01 - min(df$value[df$counts>0])
  df$value[df$counts==0] = 0
  p = ggplot(df, aes(x = x, y = y, color = value))+
    geom_point() + labs(title = gene, x = paste(reduction, 1, sep = "_"), y = paste(reduction, 2, sep = "_")) +
    scale_color_gradientn(name = "Scaled\nexpression\nlevel",
                          values = c(0, quantile(df$value[df$value>0]))/max(df$value),
                          colors = alpha(c("gray", rev(hcl.colors(5, "RdYlBu"))),0.5))+
    theme_classic()
  return(p)
}


for(EL in names(scRNA.assays)){
  x = list()
  x[["clust"]] = DimPlot(scRNA.assays[[EL]], label = TRUE) + NoLegend()
  for(el in ig){
    if(el %in% rownames(scRNA.assays[[EL]])){
      x[[el]] = gexp(scRNA = scRNA.assays[[EL]], gene = el, reduction = "UMAP")
    }
  }
  assign(EL, x)
}
```






```{r}
scRNA.assays$BT_S6$prim_ident = ifelse(scRNA.assays$BT_S6$seurat_clusters %in% c("5"), "vascular",  ifelse(scRNA.assays$BT_S6$seurat_clusters %in% c("1", "2", "4"), "immune", ifelse(scRNA.assays$BT_S6$Cell_type == "Astocyte", "astrocyte", ifelse(scRNA.assays$BT_S6$Cell_type == "Oligodendrocyte", "oligodendroglia",ifelse(scRNA.assays$BT_S6$Cell_type == "OPC", "OPC", "glioma")))))

DimPlot(scRNA.assays$BT_S6, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$BT_S4$prim_ident =  ifelse(scRNA.assays$BT_S4@reductions$umap@cell.embeddings[,"UMAP_1"]<(-10), "oligodendroglia",ifelse(scRNA.assays$BT_S4$seurat_clusters %in% c("2"), "OPC",  ifelse(scRNA.assays$BT_S4$seurat_clusters %in% c("9"), "astrocyte", ifelse(scRNA.assays$BT_S4$seurat_clusters %in% c("3"), "glioma", "immune"))))

DimPlot(scRNA.assays$BT_S4, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$BT_S1$prim_ident = ifelse(scRNA.assays$BT_S1$seurat_clusters %in% c("6"), "neuron", ifelse(scRNA.assays$BT_S1$seurat_clusters %in% c("7"), "vascular", ifelse(scRNA.assays$BT_S1$seurat_clusters %in% c("3"), "oligodendroglia", ifelse(scRNA.assays$BT_S1$seurat_clusters %in% c("2"), "astrocyte", ifelse(scRNA.assays$BT_S1$seurat_clusters %in% c("1", "5"), "immune", "glioma")))))

DimPlot(scRNA.assays$BT_S1, group.by = "prim_ident", label = T)
```

```{r}
scRNA.assays$BT_S2$prim_ident = ifelse(scRNA.assays$BT_S2$seurat_clusters %in% c("10"), "neuron", ifelse(scRNA.assays$BT_S2$seurat_clusters %in% c("2"), "OPC", ifelse(scRNA.assays$BT_S2$seurat_clusters %in% c("9"), "oligodendroglia", ifelse(scRNA.assays$BT_S2$seurat_clusters %in% c("8"), "astrocyte", ifelse(scRNA.assays$BT_S2$seurat_clusters %in% c("1", "3", "7"), "immune", "glioma")))))

DimPlot(scRNA.assays$BT_S2, group.by = "prim_ident", label = T)
```







```{r}
DimPlot(scRNA.assays$BT_S2, group.by = "prim_ident")
DimPlot(scRNA.assays$BT_S1, group.by = "prim_ident")
DimPlot(scRNA.assays$BT_S4, group.by = "prim_ident")
DimPlot(scRNA.assays$BT_S6, group.by = "prim_ident")
```


```{r}
DimPlot(scRNA.assays$BT_S2, group.by = "Cell_type")
DimPlot(scRNA.assays$BT_S1, group.by = "Cell_type")
DimPlot(scRNA.assays$BT_S4, group.by = "Cell_type")
DimPlot(scRNA.assays$BT_S6, group.by = "Cell_type")
```

```{r}
GSE84465 = scRNA.assays
save(GSE84465, file = "GSE84465.Rdata")
```

```{r}
library(xlsx)
markers = read.xlsx(file = "Gene_markers.xlsx", sheetName = "Sheet1")
markers$gene = toupper(markers$gene)

#markers = markers[markers$ofi=="yes",]
el = as.character(markers$group)
el = sapply(el, function(x){
  paste(toupper(substr(x,1,1)), substr(x,2,nchar(x)), sep = "")
})

el = sapply(el, function(x){
  if(nchar(x)>25){
    x = paste(substr(x, 1, which(strsplit(x, "")[[1]]==" ")[which.min(abs((2*which(strsplit(x, "")[[1]]==" ")/nchar(x))-1))]-1),substr(x, which(strsplit(x, "")[[1]]==" ")[which.min(abs((2*which(strsplit(x, "")[[1]]==" ")/nchar(x))-1))]+1, nchar(x)), sep = "\n")
  }
  return(x)
})
markers$group = as.factor(el)
#markers
```

```{r}
ig = markers$gene

x = list()

for(X in names(scRNA.assays)){
  el = scRNA.assays[[X]]
  x[[X]] = rowSums(el@assays$RNA@counts>0) / dim(el@assays$RNA@counts)[2]
}


x = sapply(x, function(el){
  el[unique(unlist(sapply(x, names)))]
})


x[is.na(x)] = 0


ig = intersect(ig, names(rowMeans(x)[rowMeans(x)>0.1]))
ig

save(ig, file = "ig.Rdata")
```


```{r}
#memory.limit(10*16257)
#options(future.globals.maxSize = 8*500*1024^2)


rm(list = c("BT_S1", "BT_S2", "BT_S4", "BT_S6"))

#EL <- SelectIntegrationFeatures(object.list = scRNA.assays, nfeatures = 3000)

#scRNA <- PrepSCTIntegration(object.list = scRNA.assays, anchor.features = EL)


scRNA <- FindIntegrationAnchors(object.list = scRNA.assays, anchor.features = SelectIntegrationFeatures(object.list = scRNA.assays), reference = which(names(scRNA.assays) == "BT_S1"))

rm(list = c("scRNA.assays"))

scRNA <- IntegrateData(anchorset = scRNA, features.to.integrate = unique(c(scRNA@anchor.features, ig)))
```


```{r}
#save(scRNA, file = "GSE84465_integrated_SCT.Rdata")
```


```{r}
dim = 100

scRNA = ScaleData(scRNA)
scRNA = RunPCA(scRNA, features = VariableFeatures(scRNA), npcs = dim)
```


```{r}
library(URD)
x = pcaMarchenkoPastur(M = length(VariableFeatures(scRNA)), N = dim(scRNA)[2], pca.sdev = scRNA@reductions$pca@stdev, factor = 1)
sum(!x)
dims = (1:dim)[x]

scRNA = RunUMAP(scRNA, dims = dims)
```








```{r}
'scRNA <- FindNeighbors(scRNA, dims = dims, assay = "SCT", reduction = "PCA_RNA")
for(el in 1:20/10){
  scRNA <- FindClusters(scRNA, resolution = el, assay = "SCT", reduction = "PCA_RNA")
}'

dropNA = function(x){
  x = x[!is.na(x)]
  return(x)
}


el="Cell_type"
DimPlot(scRNA,group.by = el, pt.size = 1, reduction = "umap", label = T) + scale_color_manual(values = alpha(rainbow(length(unique(dropNA(scRNA[[]][,el])))), 0.5), na.value = "gray")
```

```{r}
library(CONICSmat)
```

```{r}
x = scRNA@assays$RNA@counts
x = RelativeCounts(x, scale.factor = 1e6, verbose = TRUE)
```
```{r}
x = log2((x/10)+1)
```

```{r}
patients = colnames(x)
patients = scRNA[[]][patients,"orig.ident"]

unique(patients)
```
```{r}
regions=read.table("chromosome_arm_positions_grch38.txt",sep="\t",row.names = 1,header = T)
head(regions,n=5)
```
```{r}
gene_pos=getGenePositions(rownames(x))
```

```{r}
x=filterMatrix(x,gene_pos[,"hgnc_symbol"],minCells=20)
```

```{r}
normFactor=calcNormFactors(x)
```

```{r}
wd = getwd()
setwd(paste0(wd, "/infercnv"))
l=plotAll(x,normFactor,regions,gene_pos,"CNVs")
```

```{r}
hi=plotHistogram(l,x,clusters=5,zscoreThreshold=4,patients)
```




```{r}
ts=calculateTsne(x, intersect(rownames(x), VariableFeatures(scRNA)))
```

```{r}
umap = list()
umap$Y = scRNA@reductions$umap@cell.embeddings
```

```{r}
plotTsneProbabilities(umap,x,l[,"7p"],"Tsne Chr7p")
```

```{r}
plotTsneProbabilities(umap,x,l[,"7q"],"Tsne Chr7q")
```
```{r}
setwd(paste0(wd, "/infercnv"))

lrbic=read.table("CNVs_BIC_LR.txt",sep="\t",header=T,row.names=1,check.names=F)
colnames(lrbic)
candRegions=rownames(lrbic)[which(lrbic[,"BIC difference"]>1000 & lrbic[,"LRT adj. p-val"]<0.01)]
length(candRegions)
```
```{r}
hi=plotHistogram(l[,candRegions],x,clusters=3,zscoreThreshold=4,patients)
```

```{r}
normal= which(hi==2)
tumor=which(hi!=2)
```

```{r}
setwd(paste0(wd, "/infercnv"))
redu=plotAll(x,normFactor,regions[candRegions,],gene_pos,"CNVs_with_info.pdf",normal=normal,tumor=tumor)
```


```{r}
r=generatePvalMat(x,regions[candRegions,],normFactor,normal,tumor,gene_pos,threshold=0.8)
binr=ifelse(r>0.1,0,1)
boxplot(r)
```
```{r}
plotChromosomeHeatmap(x, normal = normal, plotcells = which(patients=="BT_S2"), gene_pos = gene_pos, windowsize = 121, chr=T, expThresh=0.2, thresh = 1)
```


```{r}
bin_mat=binarizeMatrix(redu,normal,tumor,0.8)
plotBinaryMat(bin_mat,patients,normal,tumor, patient = "BT_S2")
```


```{r}
x = c(,,,)
library(biomaRt)
mart <- useMart("ENSEMBL_MART_ENSEMBL", host = "ensembl.org")
ensembl = useDataset("hsapiens_gene_ensembl",mart=mart)

trl = getBM(attributes=c("ensembl_gene_id","hgnc_symbol", "entrezgene_id", "start_position", "end_position", "chromosome_name", "band", "strand", "description","phenotype_description", "mim_morbid_description"), mart= ensembl)
trl$chromosome_name[nchar(trl$chromosome_name)>3] = gsub("_", "", substr(trl$chromosome_name[nchar(trl$chromosome_name)>3], 10, 11))
```
```{r}

```





```{r}

scRNA$experiment = "GSE84465"
wd = getwd()
set.seed(2021)
for(smpl in unique(scRNA$orig.ident)){ 

  setwd(paste0(wd, "/infercnv"))
  
  if (file.exists(smpl)){
      setwd(paste0(wd, "/infercnv/", smpl))
  } else {
      dir.create(paste0(wd, "/infercnv/", smpl))
      setwd(paste0(wd, "/infercnv/", smpl))
  
  }
  
  scRNA$sel = ifelse(scRNA$orig.ident == smpl, 1, 0)
  
  sum(scRNA$sel)
  
  if(sum(scRNA$Cell_type[scRNA$sel == 1] == "Immune cell") < 20){
    x = table(scRNA[[]][scRNA$experiment == scRNA$experiment[match(smpl, scRNA[[]]$orig.ident)] & scRNA$Cell_type == "Immune cell","orig.ident"])
    
    scRNA$sel[sample((1:length(scRNA$orig.ident))[scRNA$Cell_type == "Immune cell" & scRNA$orig.ident == sample(names(x[x>=20]), 1)], 20)] = 1
  }
  
  sum(scRNA$sel)
  
  if(sum(scRNA$Cell_type[scRNA$sel == 1] == "Oligodendrocyte") < 20){
        x = table(scRNA[[]][scRNA$experiment == scRNA$experiment[match(smpl, scRNA[[]]$orig.ident)] & scRNA$Cell_type == "Oligodendrocyte","orig.ident"])
    
    scRNA$sel[sample((1:length(scRNA$orig.ident))[scRNA$Cell_type == "Oligodendrocyte" & scRNA$orig.ident == sample(names(x[x>=20]), 1)], 20)] = 1
  }
  
  sum(scRNA$sel)
  
  #x = table(scRNA[[]][scRNA$sel == 1,"Cell_type"])[match(scRNA$Cell_type, names(table(scRNA[[]][scRNA$sel == 1,"Cell_type"])))]
  #x[is.na(x)] = 0

  #scRNA$sel[x<2] = 0
  
  #sum(scRNA$sel)
  
  library(infercnv)
  
  infercnv_obj = subset(scRNA, sel == 1)
  
  
  X = infercnv_obj@assays$RNA@counts
  #X = as.sparse(10*((2^(X))-1))
  X = X[!is.na(match(rownames(X), trl$hgnc_symbol)),]
  
  x = trl[match(rownames(X), trl$hgnc_symbol),]
  x$chr = paste0("chr", x$chromosome_name)
  x$x = gsub("X", "23", x$chromosome_name)
  x$x = gsub("Y", "24", x$chromosome_name)
  x$x = as.numeric(x$x)
  x = x[!x$x==0,]
  x = x[!is.na(x$x),]
  x$x = x$start_position + x$x*max(x$start_position)
  x = x[order(x$x, decreasing = F),]
  
  
  x = x[x$hgnc_symbol %in% rownames(X),]
  X = X[rownames(X) %in% x$hgnc_symbol ,]
  
  dim(x)
  dim(X)
  
  
  
  rownames(x) = x$hgnc_symbol
  x = x[,c("chromosome_name", "start_position", "end_position")]
  colnames(x) = c("V2", "V3", "V4")
  
  
  
  
  el = paste(infercnv_obj$Cell_type, infercnv_obj$orig.ident, sep = "_")
  
  for(EL in c("Neuron","Vascular","Astocyte")){
    el = gsub(EL, "Neoplastic", el)
  }
  
  el = data.frame(row.names = rownames(infercnv_obj[[]]), V2 = el)
  
  print(table(el$V2))
  
  library(infercnv)
  infercnv_obj = CreateInfercnvObject(
    raw_counts_matrix=X,
    annotations_file=el,
    gene_order_file=x,
    ref_group_names= unique(el$V2[!(grepl("OPC", el$V2) | grepl("Neoplastic", el$V2))]))
  
  save(infercnv_obj, file = "infercnv_obj.Rdata")
}
```


```{r}

```




```{r}
'wd = "D:/Files/Main_directory/Nencki/Epilepsy_2020"
library(parallel)
library(infercnv)
for(smpl in rev(c("MGH60", "MGH54", "MGH53", "MGH36", "MGH42", "MGH61", "MGH43", "MGH45", "MGH44", "MGH56", "MGH64", "MGH57", "MGH107", "MGH102", "MGH104", "MGH105", "MGH110", "MGH113", "MGH121", "MGH122", "MGH124", "MGH125", "MGH136", "MGH143", "MGH66"))){
  print(smpl)
  setwd(paste0(wd, "/infercnv/", smpl))
  load(file = "infercnv_obj.Rdata")
  
  infercnv_obj = infercnv::run(
    infercnv_obj,
    cutoff=0.75, # cutoff=1 works well for Smart-seq2, and cutoff=0.1 works well for 10x Genomics
    out_dir=paste0(wd, "/infercnv/", smpl),
    cluster_by_groups=TRUE, 
    plot_steps=FALSE,
    denoise=TRUE,
    HMM=T,
    analysis_mode = "subclusters",
    tumor_subcluster_partition_method = "qnorm",
    no_prelim_plot=TRUE,
    HMM_report_by = "cell",
    png_res=300
  )
  
  save(infercnv_obj, file = "infercnv_res.Rdata")
  rm(list = "infercnv_obj")
  gc()

}'
```

```{r}
#SCT
```


```{r}
library(Seurat)
load(file = "scRNA.assays.Rdata")
```


```{r}
scRNA.assays

for(el in names(scRNA.assays)){
  x = scRNA.assays[[el]]@assays$RNA@counts
  x = RelativeCounts(x, scale.factor = 1e6, verbose = TRUE)
  scRNA.assays[[el]]@assays$RNA@counts = x
  scRNA.assays[[el]]@assays$RNA@data = x
  print(el)
}

scRNA.assays = lapply(scRNA.assays, function(mscRNA){
  mscRNA <- subset(mscRNA, subset = Doublet == "no" & percent.mt < 10 & nFeature_RNA > 400)
  return(mscRNA)
})

scRNA.assays=scRNA.assays[sapply(scRNA.assays, function(x){dim(x@assays$RNA@meta.features)[1]})>10000]
scRNA.assays=scRNA.assays[sapply(scRNA.assays, function(x){length(x$orig.ident)})>200]
print("***")
scRNA.assays
```





```{r}
#scRNA.assays = scRNA.assays[c(1,3,4)]
for (i in 1:length(scRNA.assays)) {
  scRNA.assays[[i]] <- CellCycleScoring(scRNA.assays[[i]], s.features = cc.genes.updated.2019$s.genes, g2m.features = cc.genes.updated.2019$g2m.genes, set.ident = F)
  scRNA.assays[[i]] <- SCTransform(scRNA.assays[[i]], verbose = T, return.only.var.genes = F, vars.to.regress = c("percent.mt", "S.Score", "G2M.Score"), method = 'glmGamPoi')
  print(i)
}
```




```{r}
library(URD)
```


```{r}
D = list()
p = list()
dim = 50
for (i in 1:length(scRNA.assays)) {
  scRNA.assays[[i]] <- RunPCA(scRNA.assays[[i]], verbose = FALSE, npcs = dim)
  
  dims = pcaMarchenkoPastur(M = length(VariableFeatures(scRNA.assays[[i]])), N = dim(scRNA.assays[[i]])[2], pca.sdev = scRNA.assays[[i]]@reductions$pca@stdev, factor = 1)
  dims = (1:dim)[dims]
  
  scRNA.assays[[i]] <- RunUMAP(scRNA.assays[[i]], dims = dims, verbose = FALSE)
  
  scRNA.assays[[i]] <- FindNeighbors(scRNA.assays[[i]], dims = dims, verbose = FALSE)
  scRNA.assays[[i]] <- FindClusters(scRNA.assays[[i]], verbose = FALSE)
  
  D[[i]] = dims
  p[[i]] = DimPlot(scRNA.assays[[i]], label = TRUE) + NoLegend()
}

p
```

```{r}
gexp = function(scRNA = scRNA, gene, reduction = "UMAP"){
  df = data.frame(scale = scRNA@assays$SCT@scale.data[gene,rownames(scRNA[[]])],
                  counts = scRNA@assays$RNA@counts[gene,rownames(scRNA[[]])],
                  x = scRNA@reductions[[tolower(reduction)]]@cell.embeddings[rownames(scRNA[[]]),1],
                  y = scRNA@reductions[[tolower(reduction)]]@cell.embeddings[rownames(scRNA[[]]),2])
  
  df$value = df$scale
  df$value = df$value + 0.01 - min(df$value[df$counts>0])
  df$value[df$counts==0] = 0
  p = ggplot(df, aes(x = x, y = y, color = value))+
    geom_point() + labs(title = gene, x = paste(reduction, 1, sep = "_"), y = paste(reduction, 2, sep = "_")) +
    scale_color_gradientn(name = "Scaled\nexpression\nlevel",
                          values = c(0, quantile(df$value[df$value>0]))/max(df$value),
                          colors = alpha(c("gray", rev(hcl.colors(5, "RdYlBu"))),0.5))+
    theme_classic()
  return(p)
}
```


```{r}
p = list()
pCD45 = list()
pMBP = list()
pCLU = list()
pAPOLD1 = list()
pSNAP25 = list()
pGPR17 = list()
pCD24 = list()
for (i in 1:length(scRNA.assays)) {
  p[[i]] = DimPlot(scRNA.assays[[i]], label = TRUE) + NoLegend()
  pCD45[[i]] = gexp(scRNA = scRNA.assays[[i]], gene = "PTPRC", reduction = "UMAP")
  pMBP[[i]] = gexp(scRNA = scRNA.assays[[i]], gene = "MBP", reduction = "UMAP")
  pCLU[[i]] = gexp(scRNA = scRNA.assays[[i]], gene = "CLU", reduction = "UMAP")
  pAPOLD1[[i]] = gexp(scRNA = scRNA.assays[[i]], gene = "APOLD1", reduction = "UMAP")
  pSNAP25[[i]] = gexp(scRNA = scRNA.assays[[i]], gene = "SNAP25", reduction = "UMAP")
  pGPR17[[i]] = gexp(scRNA = scRNA.assays[[i]], gene = "GPR17", reduction = "UMAP")
  pCD24[[i]] = gexp(scRNA = scRNA.assays[[i]], gene = "CD24", reduction = "UMAP")
  #scRNA.assays[[i]][[]] = scRNA.assays[[i]][[]][,!colnames(scRNA.assays[[i]][[]]) == "prim_ident"]
}
p
```






```{r}
scRNA.assays$BT_S2$prim_ident = ifelse(scRNA.assays$BT_S2$seurat_clusters %in% c("2"), "OPC", ifelse(scRNA.assays$BT_S2$seurat_clusters  %in% c("1", "0"), "immune", ifelse(scRNA.assays$BT_S2$seurat_clusters  %in% c("8"), "neuron", ifelse(scRNA.assays$BT_S2$seurat_clusters  %in% c("9"), "oligodendroglia", ifelse(scRNA.assays$BT_S2$seurat_clusters  %in% c("7"), "astrocyte", "glioma")))))
```

```{r}
scRNA.assays$BT_S1$prim_ident = ifelse(scRNA.assays$BT_S1$seurat_clusters == "4", "vascular",  ifelse(scRNA.assays$BT_S1$seurat_clusters == "3", "oligodendroglia", ifelse(scRNA.assays$BT_S1$seurat_clusters %in% c("2", "5", "6"), "immune", "glioma")))
```

```{r}
scRNA.assays$BT_S4$prim_ident = ifelse(scRNA.assays$BT_S4$Cell_type == "Oligodendrocyte", "oligodendroglia", ifelse(scRNA.assays$BT_S4$Cell_type == "Vascular", "vascular", ifelse(scRNA.assays$BT_S4$seurat_clusters %in% c("9"), "astrocyte", ifelse(scRNA.assays$BT_S4$seurat_clusters == "2", "OPC", ifelse(scRNA.assays$BT_S4$seurat_clusters %in% c("8", "7"), "glioma","immune")))))
```

```{r}
scRNA.assays$BT_S6$prim_ident = ifelse(scRNA.assays$BT_S6$seurat_clusters %in% c("3","2", "4"), "immune", ifelse(scRNA.assays$BT_S6$Cell_type == "Vascular", "vascular", ifelse(scRNA.assays$BT_S6$Cell_type == "Oligodendrocyte", "oligodendroglia",ifelse(scRNA.assays$BT_S6$Cell_type == "OPC", "OPC", ifelse(scRNA.assays$BT_S6$Cell_type == "Astocyte", "astrocyte", "glioma")))))
```











```{r}
DimPlot(scRNA.assays$BT_S2, group.by = "prim_ident")
DimPlot(scRNA.assays$BT_S1, group.by = "prim_ident")
DimPlot(scRNA.assays$BT_S4, group.by = "prim_ident")
DimPlot(scRNA.assays$BT_S6, group.by = "prim_ident")
```


```{r}
DimPlot(scRNA.assays$BT_S2, group.by = "Cell_type")
DimPlot(scRNA.assays$BT_S1, group.by = "Cell_type")
DimPlot(scRNA.assays$BT_S4, group.by = "Cell_type")
DimPlot(scRNA.assays$BT_S6, group.by = "Cell_type")
```



```{r}
#x = c(,,,)
GSE84465 = scRNA.assays
save(GSE84465, file = "GSE84465_SCT.Rdata")
```









Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
